module/VuFind/src/VuFind/Recommend/AbstractSearchObjectDeferred.php

```diff
@@ -0,0 +1,146 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Recommendations
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:recommendation_modules Wiki
+ */
+
+namespace VuFind\Recommend;
+
+/**
+ * Abstract SearchObjectDeferred Recommendations Module (needs to be extended to use
+ * a particular search object).
+ *
+ * This class sets up an AJAX call to trigger a call to some SearchObject implementation.
+ *
+ * @category VuFind
+ * @package  Recommendations
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:recommendation_modules Wiki
+ */
+abstract class AbstractSearchObjectDeferred implements RecommendInterface
+{
+    /**
+     * Raw configuration parameters
+     *
+     * @var string
+     */
+    protected $rawParams;
+
+    /**
+     * Current search query
+     *
+     * @var string
+     */
+    protected $lookfor;
+
+    /**
+     * Configuration parameters processed for submission via AJAX
+     *
+     * @var string
+     */
+    protected $processedParams;
+
+    /**
+     * Store the configuration of the recommendation module.
+     *
+     * @return string Module name in call to AjaxHandler
+     */
+    abstract protected function getAjaxModule();
+
+    /**
+     * Store the configuration of the recommendation module.
+     *
+     * @param string $settings Settings from searches.ini.
+     *
+     * @return void
+     */
+    public function setConfig($settings)
+    {
+        $this->rawParams = $settings;
+    }
+
+    /**
+     * Called before the Search Results object performs its main search
+     * (specifically, in response to \VuFind\Search\SearchRunner::EVENT_CONFIGURED).
+     * This method is responsible for setting search parameters needed by the
+     * recommendation module and for reading any existing search parameters that may
+     * be needed.
+     *
+     * @param \VuFind\Search\Base\Params $params  Search parameter object
+     * @param \Laminas\Stdlib\Parameters $request Parameter object representing user
+     * request.
+     *
+     * @return void
+     */
+    public function init($params, $request)
+    {
+        // Parse out parameters:
+        $settings = explode(':', $this->rawParams);
+
+        // Make sure all elements of the params array are filled in, even if just
+        // with a blank string, so we can rebuild the parameters to pass through
+        // AJAX later on!
+        for ($i = 0; $i < 2; $i++) {
```

Comment:
Ok, I refactored $paramCount.  And at that point it just made sense to refactor init() as well, although feedback welcome on how I did it.  

I also don't love that $paramCount is an overridden parameter and getAjaxModule is an abstract method, could use either approach for both.  I preferred the abstract method for getAjaxModule() since it forces an implementation and there is no default.  For $paramCount there is a default, but it could still be a method if that consistency is helpful.  But maybe better as is.

Reply:
I think it's okay to use different approaches for these two parameters for the reasons you state: one has a default value, and the other is required. I think it's fine as-is!